code/FormSchemaController.php

```diff
@@ -0,0 +1,148 @@
+use SilverStripe\Forms\HTMLEditor\TinyMCEConfig;
+use SilverStripe\Forms\Schema\FormSchema;
+use SilverStripe\Security\Security;
+
+/**
+ * The base class for controllers in the CMS that need access to form schema functionality.
+ */
+abstract class FormSchemaController extends AdminController
+{
+    /**
+     * Form schema header identifier
+     */
+    public const SCHEMA_HEADER = 'X-Formschema-Request';
+
+    private static array $allowed_actions = [
+        'schema',
+    ];
+
+    private static array $url_handlers = [
+        'GET schema/$FormName/$ItemID/$OtherItemID' => 'schema',
+    ];
+
+    private static array $dependencies = [
+        'FormSchema' => '%$' . FormSchema::class,
+    ];
+
+    /**
+     * Current form schema helper
+     */
+    private ?FormSchema $schema = null;
+
+    protected function init()
+    {
+        $this->beforeExtending('onInit', function () {
+            // Set the members html editor config
+            if (Security::getCurrentUser()) {
+                HTMLEditorConfig::set_active_identifier(Security::getCurrentUser()->getHtmlEditorConfigForCMS());
+            }
+
+            // Set default values in the config if missing.  These things can't be defined in the config
+            // file because insufficient information exists when that is being processed
+            $htmlEditorConfig = HTMLEditorConfig::get_active();
+            $htmlEditorConfig->setOption('language', TinyMCEConfig::get_tinymce_lang());
+            $langUrl = TinyMCEConfig::get_tinymce_lang_url();
+            if ($langUrl) {
+                $htmlEditorConfig->setOption('language_url', $langUrl);
+            }
+        });
+        parent::init();
+    }
+
+    /**
+     * Get form schema helper
+     */
+    public function getFormSchema(): FormSchema
+    {
+        if (!$this->schema) {
+            $this->schema = FormSchema::singleton();
+        }
+        return $this->schema;
+    }
+
+    /**
+     * Set form schema helper for this controller
+     */
+    public function setFormSchema(FormSchema $schema): static
+    {
+        $this->schema = $schema;
+        return $this;
+    }
+
+    /**
+     * Gets a JSON schema representing the current edit form.
+     */
+    public function schema(HTTPRequest $request): HTTPResponse
+    {
+        $formName = $request->param('FormName');
+        $itemID = $request->param('ItemID');
+
+        if (!$formName) {
+            $this->jsonError(400, 'Missing request params');
+        }
+
+        // The getter can accept an ID where the main form action wouldn't
+        $formMethod = "get{$formName}";
+        if (!$this->hasMethod($formMethod)) {
+            $formMethod = $formName;
+            if (!$this->hasMethod($formMethod)) {
+                $this->jsonError(404, 'Form not found');
+            }
+        }
+
+        if (!$this->hasAction($formName)) {
+            $this->jsonError(401, 'Form not accessible');
```

Comment:
```suggestion
            $this->jsonError(404, 'Form not accessible');
```
It's an invalid route, rather than an issue with a user not being logged in

Reply:
Changing error codes is not in scope for this PR. This isn't new code.